### ISSUE TITLE
🚀[Fine-tuning] Qwen3-MoE Megatron Training Implementation and Best Practices👋

### ISSUE THREAD

**Issue by Jintao-Huang** [QwenLM/Qwen3#1278]
Hello, everyone! We are thrilled to hear about the open-source release of Qwen3 and Qwen3-MoE. The [ms-swift](https://github.com/modelscope/ms-swift) large model training framework has provided initial support for **CPT/SFT/DPO/GRPO** training for Qwen3/Qwen3-MoE. Additionally, it supports the implementation of **Megatron** training (CPT/SFT) for Qwen3/Qwen3-MoE, which is **10 times faster** on MoE models compared to the training speed achieved using the transformers library. 

For complete best practices, you can check out the details here: https://github.com/modelscope/ms-swift/issues/4030. Everyone is welcome to try out our framework! 😊

We will showcase a runnable fine-tuning demo and provide the format for custom datasets.

Before starting the fine-tuning process, please ensure that your environment is properly set up.

```bash
git clone https://github.com/modelscope/ms-swift.git
cd ms-swift
pip install -e .

pip install liger-kernel transformers -U
```

## Qwen3-8B SFT

The script for training Qwen3-8B is as follows, which can be run on the **free A10** computing resources provided by ModelScope: https://modelscope.cn/my/mynotebook

```bash
# Training GPU memory: 22GB
# You can specify `--dataset AI-ModelScope/alpaca-gpt4-data-zh` to run the experiment
CUDA_VISIBLE_DEVICES=0 \
swift sft \
    --model Qwen/Qwen3-8B \
    --train_type lora \
    --dataset '<dataset-path>' \
    --torch_dtype bfloat16 \
    --num_train_epochs 1 \
    --per_device_train_batch_size 1 \
    --per_device_eval_batch_size 1 \
    --learning_rate 1e-4 \
    --lora_rank 8 \
    --lora_alpha 32 \
    --target_modules all-linear \
    --gradient_accumulation_steps 4 \
    --eval_steps 50 \
    --save_steps 50 \
    --save_total_limit 2 \
    --logging_steps 5 \
    --max_length 2048 \
    --output_dir output \
    --warmup_ratio 0.05 \
    --dataloader_num_workers 4 \
    --packing true \
    --user_liger_kernel true
```

The format for a **custom dataset** is as follows (the `system` field is optional). Simply specify `--dataset <dataset_path>`:

For more information, refer to the custom dataset documentation: https://swift.readthedocs.io/en/latest/Customization/Custom-dataset.html
```jsonl
{"messages": [{"role": "user", "content": "Where is the capital of Zhejiang?"}, {"role": "assistant", "content": "<think>\nxxx\n</think>\n\nThe capital of Zhejiang is Hangzhou."}]}
```

## Qwen3-8B GRPO

Taking Qwen3-8B as an example, the following uses the swift framework to conduct GRPO training. For more details about GRPO, refer to the GRPO documentation: https://swift.readthedocs.io/en/latest/Instruction/GRPO.html

The AI-MO/NuminaMath-TIR dataset is used, and the accuracy function is employed to compute the model’s response accuracy reward. The following environment needs to be installed to calculate rewards:

```bash
pip install math_verify==0.5.2
```

The custom dataset format is similar to SFT, where the assistant part is optional. If using the accuracy reward, a solution column is required to compute the accuracy.

```jsonl
# llm
{"messages": [{"role": "system", "content": "You are a useful and harmless assistant"}, {"role": "user", "content": "Tell me tomorrow's weather"}]}
{"messages": [{"role": "system", "content": "You are a useful and harmless math calculator"}, {"role": "user", "content": "What is 1 + 1?"}, {"role": "assistant", "content": "It equals 2"}, {"role": "user", "content": "What about adding 1?"}]}
{"messages": [{"role": "user", "content": "What is your name?"}]}

# mllm
{"messages": [{"role": "user", "content": "<image>What is the difference between the two images?"}], "images": ["/xxx/x.jpg"]}
{"messages": [{"role": "user", "content": "<image><image>What is the difference between the two images?"}], "images": ["/xxx/y.jpg", "/xxx/z.png"]}
```

You can also train with custom reward functions or reward models. Columns in the dataset will be passed into `**kwargs` of the reward function. An example of a custom reward function can be found here: swift/examples/train/grpo/plugin/plugin.py

```bash
    --external_plugins examples/train/grpo/plugin/plugin.py \
    --reward_funcs external_math_acc external_math_format \
    --reward_model AI-ModelScope/Skywork-Reward-Llama-3.1-8B-v0.2
```

During training, we use vLLM to accelerate the sampling process. Setting num_infer_workers=8, we deploy one vLLM engine on each device to speed up the sampling process.

The training script is as follows:

```bash
# 70G*8
CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7 \
NPROC_PER_NODE=8 \
swift rlhf \
    --rlhf_type grpo \
    --model Qwen/Qwen3-8B \
    --train_type full \
    --dataset AI-MO/NuminaMath-TIR \
    --torch_dtype bfloat16 \
    --num_train_epochs 1 \
    --per_device_train_batch_size 2 \
    --per_device_eval_batch_size 2 \
    --learning_rate 1e-6 \
    --save_total_limit 2 \
    --logging_steps 5 \
    --output_dir output \
    --gradient_accumulation_steps 1 \
    --warmup_ratio 0.05 \
    --dataloader_num_workers 4 \
    --max_completion_length 4096 \
    --vllm_max_model_len 8192 \
    --reward_funcs accuracy \
    --num_generations 16 \
    --use_vllm true \
    --vllm_gpu_memory_utilization 0.4 \
    --sleep_level 1 \
    --offload_model true \
    --offload_optimizer true \
    --gc_collect_after_offload true \
    --deepspeed zero3 \
    --num_infer_workers 8 \
    --tensor_parallel_size 1 \
    --temperature 1.0 \
    --top_p 0.85 \
    --report_to wandb \
    --log_completions true \
    --overlong_filter true 
```

## Qwen3-30B-A3B MoE SFT (Megatron-SWIFT)

SWIFT introduces Megatron's parallel technology to accelerate large model training, including data parallelism, tensor parallelism, pipeline parallelism, sequence parallelism, context parallelism, and expert parallelism. It supports pre-training and fine-tuning of models like **Qwen3**, **Qwen3-MoE**, Qwen2.5, Llama3, Deepseek-R1 distillation series, etc.

For environment preparation (image) and the conversion between HF and MCore model weights, please refer to the Megatron-SWIFT training documentation; it is not covered here: https://swift.readthedocs.io/en/latest/Instruction/Megatron-SWIFT-Training.html

We use DLC to initiate the training command. The training environment consists of 2 machines with 8 * 80GiB A800:

More **multi-node** launch methods can be found here: https://github.com/modelscope/ms-swift/tree/main/examples/train/multi-node

```bash
# https://help.aliyun.com/zh/pai/user-guide/general-environment-variables
# Please ensure that the weight saving paths are the same for both nodes.
NNODES=$WORLD_SIZE \
NODE_RANK=$RANK \
megatron sft \
    --load Qwen3-30B-A3B-Base-mcore \
    --dataset 'liucong/Chinese-DeepSeek-R1-Distill-data-110k-SFT' \
    --tensor_model_parallel_size 2 \
    --expert_model_parallel_size 8 \
    --moe_grouped_gemm true \
    --moe_shared_expert_overlap true \
    --moe_aux_loss_coeff 0.01 \
    --micro_batch_size 1 \
    --global_batch_size 16 \
    --packing true \
    --recompute_granularity full \
    --recompute_method uniform \
    --recompute_num_layers 1 \
    --train_iters 2000 \
    --eval_iters 50 \
    --finetune true \
    --cross_entropy_loss_fusion true \
    --lr 1e-5 \
    --lr_warmup_iters 100 \
    --min_lr 1e-6 \
    --save megatron_output/Qwen3-30B-A3B-Base \
    --eval_interval 200 \
    --save_interval 200 \
    --max_length 8192 \
    --num_workers 8 \
    --dataset_num_proc 8 \
    --no_save_optim true \
    --no_save_rng true \
    --sequence_parallel true \
    --use_flash_attn true
```

The custom dataset format is the same as `swift sft`, which can be found above. Specify `--dataset <dataset_path>`.

Below is the comparison of full-parameter training **speed/GPU memory** usage for the Qwen3-30B-A3B model using `megatron sft` and `swift sft`:

|                  | Megatron-LM | DeepSpeed-ZERO2 | DeepSpeed-ZERO3 |
| ---------------- | ----------- | --------------- | --------------- |
| Training Speed   | 9.6s/it     | -               | 91.2s/it        |
| GPU Memory Usage | 16 * 60GiB  | OOM             | 16 * 80GiB      |

**Comment by yurochang** [comment]
could the GRPO work on 910b?

**Comment by cwz427** [comment]
> could the GRPO work on 910b?

大概率可以，没实践。

**Comment by rookiebird** [comment]
牛的，俺来试试， 之前试过开 deepspeed zero3 训练 qwen-moe-a2.7B 的那个模型，很早之前开源的， 发现对比同尺寸的14B 的模型，训练速度竟然还要更慢， 都是单机8 卡的模型， 一直怀疑是不是自己MOE 的配置有问题

**Comment by jklj077** [comment]
@Jintao-Huang Thanks for the guide on finetuneing! It would be great if you could also update the ms-swift documentation at <https://github.com/QwenLM/Qwen3/blob/main/docs/source/training/ms_swift.rst>.

**Comment by Jintao-Huang** [comment]
> > could the GRPO work on 910b?
> 
> 大概率可以，没实践。

哈哈哈哈，支持的

**Comment by Jintao-Huang** [comment]
> [@Jintao-Huang](https://github.com/Jintao-Huang) Thanks for the guide on finetuneing! It would be great if you could also update the ms-swift documentation at https://github.com/QwenLM/Qwen3/blob/main/docs/source/training/ms_swift.rst.

Alright, thanks ~
